### PR TITLE
CPP: Improve  ReturnStackAllocatedMemory.ql

### PR DESCRIPTION
**Pull request by geoffw0**
Improve `ReturnStackAllocatedMemory.ql`, and add a test.  This query will now catch line 43 of https://github.com/gsingh93/lgtm-test/blob/master/test.cpp.

I ran this on 38 projects on LGTM (https://lgtm.com/query/3686583108945505816/).  Most results were false positives due to poor dataflow, along the lines of:
```
MyClass local;
int ptr = &local;

...

ptr = new MyClass;

...

return ptr;
```
If we fix that at some point we could put this query on LGTM, though in any case it won't produce large numbers of results.

### CODE REVIEW

**Comment by geoffw0** [comment]
Fixed merge conflict.

**Comment by rdmarsh2** [comment]
I'm curious whether we'd do better with `DataFlow::localFlow` in either the AST or IR version.

File: cpp/ql/test/query-tests/Likely Bugs/Memory Management/ReturnStackAllocatedMemory/test.cpp [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+	return ptr; // BAD
+}
+
+MyClass *test3()
+{
+	MyClass mc;
+	MyClass &ref = mc;
+
+	return &ref; // BAD [NOT DETECTED]
+}
```
Comment (jbj) [inline cpp/ql/test/query-tests/Likely Bugs/Memory Management/ReturnStackAllocatedMemory/test.cpp]: A related test case we might add:

```
MyClass &test3b() {
    MyClass mc;
    return mc;
}
```

File: cpp/ql/test/query-tests/Likely Bugs/Memory Management/ReturnStackAllocatedMemory/test.cpp [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+	return &mc; // BAD
+}
+
+MyClass *test2()
+{
+	MyClass mc;
+	MyClass *ptr = &mc;
+
+	return ptr; // BAD
+}
```
Comment (jbj) [inline cpp/ql/test/query-tests/Likely Bugs/Memory Management/ReturnStackAllocatedMemory/test.cpp]: I'd like to see these test cases too:

```
MyClass *test2b()
{
	MyClass mc;
	MyClass *ptr = &mc;
	ptr = nullptr;

	return ptr; // GOOD
}

MyClass makeMyClass() {
    return { 0, 0 }; // GOOD
}
```

**Comment by geoffw0** [comment]
> I'm curious whether we'd do better with DataFlow::localFlow in either the AST or IR version.

Me too, the difficulty being that this kind of upgrade often requires a lot of testing and iteration. 
 Fortunately this particular query isn't used in the samate tests, so there's one less thing to think about.

**Comment by geoffw0** [comment]
Note to self: @jbj is looking into some improvements to this query / PR.

File: cpp/ql/test/query-tests/Likely Bugs/Memory Management/ReturnStackAllocatedMemory/ReturnCstrOfLocalStdString.qlref [hunk 3]
```diff
@@ -0,0 +1 @@
+Likely Bugs/Memory Management/ReturnStackAllocatedMemory.ql
```
[the remark this answers is not in the document]
Comment (geoffw0) [reply cpp/ql/test/query-tests/Likely Bugs/Memory Management/ReturnStackAllocatedMemory/ReturnCstrOfLocalStdString.qlref]: Whoops, copy-paste error.  Now fixed.

File: cpp/ql/test/query-tests/Likely Bugs/Memory Management/ReturnStackAllocatedMemory/test.cpp (hunk 1 above)
Comment (geoffw0) [reply cpp/ql/test/query-tests/Likely Bugs/Memory Management/ReturnStackAllocatedMemory/test.cpp]: Added.

File: cpp/ql/test/query-tests/Likely Bugs/Memory Management/ReturnStackAllocatedMemory/test.cpp (hunk 2 above)
Comment (geoffw0) [reply cpp/ql/test/query-tests/Likely Bugs/Memory Management/ReturnStackAllocatedMemory/test.cpp]: Added.

**Comment by jbj** [comment]
I've opened a PR against your PR (https://github.com/geoffw0/ql/pull/1) with the changes I had in mind.

**Comment by geoffw0** [comment]
This should be ready to merge now.  I mentioned spotting a typo in your PR @jbj, but it turns out you had the correct American spelling so there's nothing to correct.  :+1:

**Pull request merged**